### ISSUE TITLE
Classification Threshold

### ISSUE THREAD

**Issue by hzphzp** [facebookresearch/detr#201]
## ❓ Do you use a classification threshold?

I notice that you input all [query_size X 4] prediction bounding box to evaluator:
1. why don't use a class threshold to remove Nonetype bbox prediction.
2. Or are there any related codes I have not seen?

```
        out_logits, out_bbox = outputs['pred_logits'], outputs['pred_boxes']
        assert len(out_logits) == len(target_sizes)
        assert target_sizes.shape[1] == 2

        prob = F.softmax(out_logits, -1)
        scores, labels = prob[..., :-1].max(-1)

        # convert to [x0, y0, x1, y1] format
        boxes = box_ops.box_cxcywh_to_xyxy(out_bbox)
        # and from relative [0, 1] to absolute [0, height] coordinates
        img_h, img_w = target_sizes.unbind(1)
        scale_fct = torch.stack([img_w, img_h, img_w, img_h], dim=1)
        boxes = boxes * scale_fct[:, None, :]

        results = [{'scores': s, 'labels': l, 'boxes': b} for s, l, b in zip(scores, labels, boxes)]

        return results
```

**Comment by alcinos** [comment]
Hi @hzphzp 
For evaluation we do not use any threshold. Relevant quote from the paper: 

> "At inference time, some slots predict empty class. To optimize for AP, we override the prediction of these slots with the second highest scoring class, using the corresponding confidence. This improves AP by 2 points compared to filtering out empty slots."

We use a threshold for visualization though, please refer to our example colab for more information.

I believe I have answered your question, and as such I'm closing this. Feel free to reach out if you have further concerns.

**Comment by BrandaZhao** [comment]
> Hi @hzphzp
> For evaluation we do not use any threshold. Relevant quote from the paper:
> 
> > "At inference time, some slots predict empty class. To optimize for AP, we override the prediction of these slots with the second highest scoring class, using the corresponding confidence. This improves AP by 2 points compared to filtering out empty slots."
> 
> We use a threshold for visualization though, please refer to our example colab for more information.
> 
> I believe I have answered your question, and as such I'm closing this. Feel free to reach out if you have further concerns.
Hi! Could you please tell me where are codes for 'To optimize for AP, we override the prediction of these slots with the second highest scoring class, using the corresponding confidence.'? I didn't find this. THANKS!

**Comment by xldagger** [comment]
> > Hi @hzphzp
> > For evaluation we do not use any threshold. Relevant quote from the paper:
> > > "At inference time, some slots predict empty class. To optimize for AP, we override the prediction of these slots with the second highest scoring class, using the corresponding confidence. This improves AP by 2 points compared to filtering out empty slots."
> > 
> > 
> > We use a threshold for visualization though, please refer to our example colab for more information.
> > I believe I have answered your question, and as such I'm closing this. Feel free to reach out if you have further concerns.
> 
> **Hi! Could you please tell me where are codes for 'To optimize for AP, we override the prediction of these slots with the second highest scoring class, using the corresponding confidence.'? I didn't find this. THANKS!**


here,  
`scores, labels = prob[..., **:-1**].max(-1)`
